networkx/algorithms/centrality/trophic.py

```diff
@@ -40,10 +40,7 @@ def trophic_levels(G, weight="weight"):
     ----------
     .. [1] Stephen Levine (1980) J. theor. Biol. 83, 195-207
     """
-    try:
-        import numpy as np
-    except ImportError as e:
-        raise ImportError("trophic_levels() requires NumPy: http://numpy.org/") from e
+    import numpy as np
```

Comment:
Is there any particular performance gain/loss by importing inside the function?
Now that the try and except are gone, maybe we should move all the `from/import numpy/scipy` to top of the files?

Reply:
If we move the imports outside the functions, we will need to add come code to avoid failures in pypy (which doesn't have numpy, scipy, etc. installed).  I am not sure what the best way to handle this, but this seemed like a good first step.  (I.e., it would be easier to see what I removed and how it was replaced.)  I am happy to revisit this, but I am not sure how best to keep pypy from complaining.  I am hoping that numpy, scipy, etc. will eventually release wheels for pypy.

Reply:
My vote would be for a tiered approach - tackle the `try/excepts` in this PR then evaluate moving imports out (which I agree is likely desirable in a lot of cases aside from the pypy support issue) in a follow-up.

Reply:
I will follow-up on this.